### PR TITLE
[CPTwentyFour] Add extractor

### PR DESCRIPTION
**Pull request by Ashish0804**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Closes https://github.com/yt-dlp/yt-dlp/issues/1768

### CODE REVIEW

File: yt_dlp/extractor/ninecninemedia.py [hunk 1]
```diff
@@ -99,3 +100,40 @@ def _real_extract(self, url):
...
+            'thumbnail': 'http://images2.9c9media.com/image_asset/2014_11_5_2eb609a0-475b-0132-fbd6-34b52f6f1279_jpg_2000x1125.jpg',
+            'duration_string': '26',
+            'upload_date': '20211122',
+        },
+        'params': {'skip_download': True, 'format': 'bv'}
+    }]
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, None)
+        match_obj = re.search(r'getAuthStates\(\"(?P<id>[^\"]+)\",\s?\"(?P<destination>[^\"]+)\"\);', webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/ninecninemedia.py:133]: why `\"` everywhere?

Comment (Ashish0804) [reply yt_dlp/extractor/ninecninemedia.py:133]: Have to escape it for regex101.
Not sure if it's needed in python or not tbh...i just copy the pattern from regex101 to code
>" An unescaped delimiter must be escaped with a backslash (\)

Comment (pukkandan) [reply yt_dlp/extractor/ninecninemedia.py:133]: u need to escape whichever quote is used for the string definition. In out case that is `'`. So `"` shouldnt be escaped

File: yt_dlp/extractor/ninecninemedia.py [hunk 2]
```diff
@@ -99,3 +100,40 @@ def _real_extract(self, url):
             }
 
         return info
+
+
+class CPTwentyFourIE(InfoExtractor):
+    IE_NAME = 'cp24'
+    _GEO_COUNTRIES = ['CA']
+    _VALID_URL = r'https?://(?:www\.)?cp24\.com/news/.*'
```
Comment (pukkandan) [inline yt_dlp/extractor/ninecninemedia.py:108]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?cp24\.com/news/(?P<id>[^?#]+)'
```
and use this id for the webpage download message
and maybe also return as `display_id`

File: yt_dlp/extractor/ninecninemedia.py [hunk 3]
```diff
@@ -99,3 +100,40 @@ def _real_extract(self, url):
...
+            'description': 'md5:cf7498480885f080a754389a2b2f7073',
+            'timestamp': 1637618377,
+            'episode_number': None,
+            'season': 'Season 0',
+            'season_number': 0,
+            'season_id': 57974,
+            'series': 'CTV News Toronto',
+            'duration': 26.86,
+            'thumbnail': 'http://images2.9c9media.com/image_asset/2014_11_5_2eb609a0-475b-0132-fbd6-34b52f6f1279_jpg_2000x1125.jpg',
+            'duration_string': '26',
```
Comment (pukkandan) [inline yt_dlp/extractor/ninecninemedia.py:125]: `duration_string` is auto-generated and shouldn't be tested

File: yt_dlp/extractor/ninecninemedia.py [hunk 4]
```diff
@@ -99,3 +100,40 @@ def _real_extract(self, url):
...
+    }]
+
+    def _real_extract(self, url):
+        webpage = self._download_webpage(url, None)
+        match_obj = re.search(r'getAuthStates\(\"(?P<id>[^\"]+)\",\s?\"(?P<destination>[^\"]+)\"\);', webpage)
+        if not match_obj:
+            raise self.raise_no_formats('No video found on the page')
+        id, destination = match_obj.groups()
+        if not id and not destination:
+            raise self.raise_no_formats('Unable to parse id or destination')
```
Comment (pukkandan) [inline yt_dlp/extractor/ninecninemedia.py:138]:
```suggestion
        video_id, destination = self._search_regex(
            r'getAuthStates\("(?P<id>[^"]+)",\s?"(?P<destination>[^"]+)"\);',
            webpage, 'video id and destination', group=('id', 'destination'))
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/ninecninemedia.py (hunk 3 above)
Comment (Ashish0804) [reply yt_dlp/extractor/ninecninemedia.py]: Done

File: yt_dlp/extractor/ninecninemedia.py (hunk 2 above)
Comment (Ashish0804) [reply yt_dlp/extractor/ninecninemedia.py]: Looks ugly tbh but okay

**Pull request merged**